src/services/services.ts

```diff
@@ -2912,6 +2919,63 @@ module ts {
+                        token.parent.kind === SyntaxKind.ExportDeclaration ||
+                        token.parent.kind === SyntaxKind.ExternalModuleReference;
+                }
+                return false;
+            }
+
+            function getExternalModulecompletions(location: Node): CompletionInfo {
+                var entries: CompletionEntry[] = [];
+                var currentSourceFile = location.getSourceFile();
+                var currentDirectoryPath = getDirectoryPath(normalizePath(currentSourceFile.fileName));
+                var isInAmbientContext = ts.isInAmbientContext(location);
+
+                forEach(program.getSourceFiles(), sourceFile => {
+                    if (isExternalModule(sourceFile)) {
+                        // Only consider these if current request is not in an ambient context,
+                        // and in a diffrent file
+                        if (!isInAmbientContext && sourceFile !== currentSourceFile) {
+                            var relativeModuleName = removeFileExtension(getRelativePathToDirectoryOrUrl(currentDirectoryPath,
+                                sourceFile.fileName, "", getCanonicalFileName, /*isAbsolutePathAnUrl*/ true));
+                            if (relativeModuleName.charCodeAt(0) !== CharacterCodes.dot && relativeModuleName.charCodeAt(1) !== CharacterCodes.slash) {
+                                // always make files relative
+                                relativeModuleName = "./" + relativeModuleName;
+                            }
+                            entries.push({
+                                name: relativeModuleName,
+                                kind: ScriptElementKind.stringElement,
+                                kindModifiers: ScriptElementKindModifier.none,
+                            });
+                        }
+                    }
+                    else {
+                        // Look for ambient external module declarations
+                        forEachChild(sourceFile, child => {
+                            if (child.kind === SyntaxKind.ModuleDeclaration && (<ModuleDeclaration>child).name.kind === SyntaxKind.StringLiteral) {
+                                if (sourceFile !== currentSourceFile || location.pos > child.end || location.end < child.pos) {
+                                    entries.push({
+                                        kind: ScriptElementKind.stringElement,
+                                        kindModifiers: ScriptElementKindModifier.ambientModifier,
+                                        name: (<ModuleDeclaration>child).name.text
+                                    });
+                                }
+                            }
+                        });
+                    }
+                });
+
+                return {
+                    isMemberCompletion: true,
+                    isNewIdentifierLocation: true,
+                    isBuilder: true,

```

Comment:
Hmm, I don't understand why these are true. It certainly doesn't seem like a member completion. Also, what is the difference between isNewIdentifierLocation and isBuilder? I thought those both correspond to the notion that an entry should not be implicitly committed.